File: ui/static/assets/js/tabs/suppressions.js
import { apiGet, apiPost } from "../lib/api-client.js";
import { $, $all, clearNode, el } from "../lib/dom.js";
import { formatDate, mapStatusClass } from "../lib/format.js";

let ctx = null;

function matcherRow(defaults = {}) {
  const row = el("div", { className: "kv-pair" });

  const keyInput = el("input", {
    attrs: {
      type: "text",
      placeholder: "label key (e.g. alertname)",
      value: defaults.label_key || "",
      class: "supp-match-key",
    },
  });

  const operator = el("select", { attrs: { class: "supp-match-operator" } });
  ["eq", "neq", "regex", "nregex", "exists", "not_exists"].forEach((item) => {
    const option = el("option", { attrs: { value: item }, text: item });
    operator.appendChild(option);
  });
  operator.value = defaults.operator || "eq";

  const valueInput = el("input", {
    attrs: {
      type: "text",
      placeholder: "value (optional for exists/not_exists)",
      value: defaults.value || "",
      class: "supp-match-value",
    },
  });

  const remove = el("button", {
    className: "btn small danger",
    text: "Remove",
    on: {
      click: () => {
        row.remove();
      },
    },
  });

  row.appendChild(keyInput);
  row.appendChild(operator);
  row.appendChild(valueInput);
  row.appendChild(remove);

  return row;
}

function readMatchers() {
  return $all("#supp-matcher-list .kv-pair")
    .map((row) => {
      const labelKey = row.querySelector(".supp-match-key")?.value?.trim();
      const operator = row.querySelector(".supp-match-operator")?.value;
      const value = row.querySelector(".supp-match-value")?.value?.trim();
      return {
        label_key: labelKey,
        operator,
        value: value || null,
      };
    })
    .filter((item) => item.label_key);
}

function renderSuppressedActivity(items) {
  const container = $("#suppressed-activity");
  clearNode(container);

  if (!Array.isArray(items) || !items.length) {
    container.appendChild(el("p", { className: "empty", text: "No suppressed activity." }));
    return;
  }

  items.forEach((item) => {
    const block = el("div", { className: "activity-item" }, [
      el("div", { className: "row between" }, [
        el("strong", { text: item.alertname || "Unknown" }),
        el("span", { className: `badge ${mapStatusClass(item.status)}`, text: item.status || "-" }),
      ]),
      el("div", { className: "muted", text: `Suppression #${item.suppression_id} | ${item.severity || "unknown"}` }),
      el("div", { className: "time", text: formatDate(item.received_at) }),
    ]);
    container.appendChild(block);
  });
}

export async function loadSuppressedActivity() {
  const data = await apiGet("/api/v1/activity/suppressed?limit=30");
  renderSuppressedActivity(data);
}

async function cancelSuppression(id) {
  const ok = await ctx.confirm(`Cancel suppression #${id}?`);
  if (!ok) return;

  await apiPost(`/api/v1/suppressions/${id}/cancel`, null);
  ctx.notify("success", `Suppression #${id} canceled`);
  await load();
  await ctx.refreshDashboard();
}

function suppressionTableRow(item, totalSuppressed = 0) {
  const tr = el("tr");
  tr.appendChild(el("td", { text: item.id }));

  const nameCell = el("td");
  nameCell.appendChild(el("strong", { text: item.name || "-" }));
  if (item.reason) {
    nameCell.appendChild(el("div", { className: "muted", text: item.reason }));
  }
  tr.appendChild(nameCell);

  tr.appendChild(el("td", {}, [
    el("span", { className: `badge ${mapStatusClass(item.status)}`, text: item.status || "unknown" }),
  ]));
  tr.appendChild(el("td", { text: item.scope || "-" }));
  tr.appendChild(el("td", { text: `${formatDate(item.starts_at)} / ${formatDate(item.ends_at)}` }));
  tr.appendChild(el("td", { text: totalSuppressed }));

  const actions = el("td");
  actions.appendChild(
    el("button", {
      className: "btn small danger",
      text: "Cancel",
      on: {
        click: () => cancelSuppression(item.id),
      },
    }),
  );
  tr.appendChild(actions);

  return tr;
}

export async function load() {
  const status = $("#suppression-status-filter")?.value || "";
  const url = status
    ? `/api/v1/suppressions?status=${encodeURIComponent(status)}`
    : "/api/v1/suppressions";

  const rows = await apiGet(url);
  const tbody = $("#supp-table-body");
  clearNode(tbody);

  if (!Array.isArray(rows) || !rows.length) {
    const tr = el("tr");
    tr.appendChild(el("td", { attrs: { colspan: "7" }, className: "empty", text: "No suppressions found" }));
    tbody.appendChild(tr);
  } else {
    for (const item of rows) {
      let totalSuppressed = 0;
      try {
        const stats = await apiGet(`/api/v1/suppressions/${item.id}/stats`);
        totalSuppressed = stats?.total_suppressed || 0;
      } catch {
        totalSuppressed = 0;
      }
      tbody.appendChild(suppressionTableRow(item, totalSuppressed));
    }
  }

  await loadSuppressedActivity();
}

export async function createSuppression() {
  const name = $("#supp-name")?.value?.trim();
  const reason = $("#supp-reason")?.value?.trim();
  const startsAt = $("#supp-starts-at")?.value?.trim();
  const endsAt = $("#supp-ends-at")?.value?.trim();
  const scope = $("#supp-scope")?.value || "matchers";
  const summaryTicketEnabled = Boolean($("#supp-summary-ticket")?.checked);

  if (!name || !startsAt || !endsAt) {
    ctx.notify("error", "Name, starts_at, and ends_at are required.");
    return;
  }

  const payload = {
    name,
    reason: reason || null,
    starts_at: startsAt,
    ends_at: endsAt,
    scope,
    created_by: "ui",
    summary_ticket_enabled: summaryTicketEnabled,
    enabled: true,
    matchers: [],
  };

  if (scope === "matchers") {
    const matchers = readMatchers();
    if (!matchers.length) {
      ctx.notify("error", "At least one matcher is required when scope=matchers.");
      return;
    }
    payload.matchers = matchers;
  }

  await apiPost("/api/v1/suppressions", payload);
  ctx.notify("success", "Suppression created");
  await load();
  await ctx.refreshDashboard();
}

export function init(context) {
  ctx = context;

  $("#supp-add-matcher")?.addEventListener("click", () => {
    $("#supp-matcher-list")?.appendChild(matcherRow());
  });

  $("#supp-create")?.addEventListener("click", () => {
    createSuppression().catch((error) => {
      ctx.notify("error", error.message);
    });
  });

  $("#suppression-refresh")?.addEventListener("click", () => {
    load().catch((error) => {
      ctx.notify("error", error.message);
    });
  });

  $("#suppression-status-filter")?.addEventListener("change", () => {
    load().catch((error) => {
      ctx.notify("error", error.message);
    });
  });

  if (!$all("#supp-matcher-list .kv-pair").length) {
    $("#supp-matcher-list")?.appendChild(matcherRow());
  }
}

File: cli/commands/suppressions.py
"""Suppression commands for the PoundCake CLI."""

from __future__ import annotations

import getpass
from typing import Any

import click

from cli.client import PoundCakeClientError
from cli.commands.common import get_client, get_output_format
from cli.utils import parse_json_object, print_error, print_output, render_sections, to_plain_data


def _suppression_rows(rows: list[dict[str, Any]]) -> list[dict[str, Any]]:
    return [
        {
            "id": item.get("id"),
            "name": item.get("name"),
            "status": item.get("status"),
            "scope": item.get("scope"),
            "enabled": item.get("enabled"),
            "starts_at": item.get("starts_at"),
            "ends_at": item.get("ends_at"),
        }
        for item in rows
    ]


def _suppression_detail_table(item: dict[str, Any]) -> str:
    sections: list[tuple[str, Any]] = [
        (
            "Suppression",
            {
                "id": item.get("id"),
                "name": item.get("name"),
                "status": item.get("status"),
                "scope": item.get("scope"),
                "enabled": item.get("enabled"),
                "starts_at": item.get("starts_at"),
                "ends_at": item.get("ends_at"),
                "reason": item.get("reason"),
                "created_by": item.get("created_by"),
                "summary_ticket_enabled": item.get("summary_ticket_enabled"),
            },
        ),
        (
            "Matchers",
            [
                {
                    "label_key": matcher.get("label_key"),
                    "operator": matcher.get("operator"),
                    "value": matcher.get("value"),
                }
                for matcher in item.get("matchers") or []
            ],
        ),
    ]
    counters = item.get("counters")
    if counters:
        sections.append(("Counters", counters))
    summary = item.get("summary")
    if summary:
        sections.append(("Summary", summary))
    return render_sections(sections)


def _build_matchers(
    matcher_key: str | None,
    matcher_operator: str | None,
    matcher_value: str | None,
    matcher_json: tuple[str, ...],
) -> list[dict[str, Any]]:
    matchers: list[dict[str, Any]] = []
    if matcher_key:
        matchers.append(
            {
                "label_key": matcher_key,
                "operator": matcher_operator or "eq",
                "value": matcher_value,
            }
        )
    for raw in matcher_json:
        matchers.append(parse_json_object(raw, "matcher-json") or {})
    return matchers


@click.group(name="suppressions")
def suppressions() -> None:
    """Manage suppression windows."""


@suppressions.command("list")
@click.option(
    "--status", type=click.Choice(["scheduled", "active", "expired", "canceled"]), default=None
)
@click.option("--enabled/--disabled", default=None)
@click.option("--scope", type=click.Choice(["all", "matchers"]), default=None)
@click.option("--limit", type=int, default=100, show_default=True)
@click.option("--offset", type=int, default=0, show_default=True)
@click.pass_context
def list_suppressions_cmd(
    ctx: click.Context,
    status: str | None,
    enabled: bool | None,
    scope: str | None,
    limit: int,
    offset: int,
) -> None:
    """List suppressions."""
    client = get_client(ctx)
    output_format = get_output_format(ctx)
    try:
        payload = client.list_suppressions(
            status=status, enabled=enabled, scope=scope, limit=limit, offset=offset
        )
        if output_format == "table":
            print_output(_suppression_rows(to_plain_data(payload)), output_format)
            return
        print_output(payload, output_format)
    except PoundCakeClientError as exc:
        print_error(f"Failed to list suppressions: {exc}")
        raise click.Abort() from exc


@suppressions.command("get")
@click.argument("suppression_id", type=int)
@click.pass_context
def get_suppression_cmd(ctx: click.Context, suppression_id: int) -> None:
    """Get one suppression."""
    client = get_client(ctx)
    output_format = get_output_format(ctx)
    try:
        payload = client.get_suppression(suppression_id)
        print_output(payload, output_format, table_renderer=_suppression_detail_table)
    except PoundCakeClientError as exc:
        print_error(f"Failed to get suppression: {exc}")
        raise click.Abort() from exc


@suppressions.command("create")
@click.option("--name", required=True)
@click.option("--starts-at", required=True, help="ISO-8601 start time")
@click.option("--ends-at", required=True, help="ISO-8601 end time")
@click.option(
    "--scope", type=click.Choice(["all", "matchers"]), default="matchers", show_default=True
)
@click.option("--reason", default=None)
@click.option("--created-by", default=None)
@click.option("--summary-ticket-enabled/--summary-ticket-disabled", default=True, show_default=True)
@click.option("--matcher-key", default=None)
@click.option("--matcher-operator", default="eq", show_default=True)
@click.option("--matcher-value", default=None)
@click.option("--matcher-json", multiple=True, help="JSON object matcher payload")
@click.pass_context
def create_suppression_cmd(
    ctx: click.Context,
    name: str,
    starts_at: str,
    ends_at: str,
    scope: str,
    reason: str | None,
    created_by: str | None,
    summary_ticket_enabled: bool,
    matcher_key: str | None,
    matcher_operator: str,
    matcher_value: str | None,
    matcher_json: tuple[str, ...],
) -> None:
    """Create a suppression window."""
    client = get_client(ctx)
    output_format = get_output_format(ctx)
    try:
        matchers = _build_matchers(matcher_key, matcher_operator, matcher_value, matcher_json)
        if scope == "matchers" and not matchers:
            raise click.BadParameter("At least one matcher is required when scope=matchers")
        payload = {
            "name": name,
            "starts_at": starts_at,
            "ends_at": ends_at,
            "scope": scope,
            "matchers": matchers,
            "reason": reason,
            "created_by": created_by or getpass.getuser() or "cli",
            "summary_ticket_enabled": summary_ticket_enabled,
            "enabled": True,
        }
        response = client.create_suppression(payload)
        print_output(response, output_format, table_renderer=_suppression_detail_table)
    except (click.BadParameter, PoundCakeClientError) as exc:
        print_error(f"Failed to create suppression: {exc}")
        raise click.Abort() from exc


@suppressions.command("cancel")
@click.argument("suppression_id", type=int)
@click.pass_context
def cancel_suppression_cmd(ctx: click.Context, suppression_id: int) -> None:
    """Cancel a suppression window."""
    client = get_client(ctx)
    output_format = get_output_format(ctx)
    try:
        payload = client.cancel_suppression(suppression_id)
        print_output(payload, output_format)
    except PoundCakeClientError as exc:
        print_error(f"Failed to cancel suppression: {exc}")
        raise click.Abort() from exc
